syft/frameworks/torch/fl/utils.py

```diff
@@ -29,60 +29,67 @@ def extract_batches_per_worker(federated_train_loader: sy.FederatedDataLoader):
     return batches
 
 
-def add_model(dst_model, src_model):
+def add_model(dst_model, src_model,dst_num_data,src_num_data):
     """Add the parameters of two models.
-
         Args:
+
             dst_model (torch.nn.Module): the model to which the src_model will be added.
             src_model (torch.nn.Module): the model to be added to dst_model.
+            dst_num_data: the number of data within dst_model.
+            src_num_data: the number of data within src_model.
+
         Returns:
+        
             torch.nn.Module: the resulting model of the addition.
-
         """
-
+    if (dst_model==None):
+        return src_model
     params1 = src_model.named_parameters()
     params2 = dst_model.named_parameters()
     dict_params2 = dict(params2)
     with torch.no_grad():
         for name1, param1 in params1:
             if name1 in dict_params2:
-                dict_params2[name1].set_(param1.data + dict_params2[name1].data)
+                dict_params2[name1].set_(param1.data*src_num_data + dict_params2[name1].data*dst_num_data)
     return dst_model
 
 
 def scale_model(model, scale):
     """Scale the parameters of a model.
-
     Args:
+
         model (torch.nn.Module): the models whose parameters will be scaled.
         scale (float): the scaling factor.
+
     Returns:
-        torch.nn.Module: the module with scaled parameters.
 
+        torch.nn.Module: the module with scaled parameters.
     """
     params = model.named_parameters()
     dict_params = dict(params)
     with torch.no_grad():
-        for name, param in dict_params.items():
+        for name, _ in dict_params.items():
             dict_params[name].set_(dict_params[name].data * scale)
     return model
 
 
-def federated_avg(models: List[torch.nn.Module]) -> torch.nn.Module:
+def federated_avg(models,data_num):
     """Calculate the federated average of a list of models.
-
     Args:
-        models (List[torch.nn.Module]): the models of which the federated average is calculated.
 
+        models: the list of models of which the federated average is calculated.
+        data_num: the list of number of data within corresponding model.
     Returns:
+
         torch.nn.Module: the module with averaged parameters.
     """
-    nr_models = len(models)
-    model_list = list(models.values())
-    model = model_list[0]
-    for i in range(1, nr_models):
-        model = add_model(model, model_list[i])
-    model = scale_model(model, 1.0 / nr_models)
+
+    total_num_data=0
+    model=None
+    for i in range(len(models)):
+        model = add_model(model, models[i],total_num_data,data_num[i])
+        model = scale_model(model, 1.0 / (total_num_data+data_num[i]))
```

Comment:
Would it not be simpler to scale the model outside the loop and in `add_model` only send the `data_num[i]` parameter? At the moment you're scaling the model by `total_num_data` and then multiplying it by `total_num_data` the next time you add a model

Reply:
I agree, this code looks correct, but the new feature you're including here might be an opportunity to refactor this file to make it more intuitive. If you're willing to do this thats would be great! Otherwise I'm afaird the next person willing to improve the features will break smthg because it doesn't fully understand what's going on.